### PR TITLE
journald: enforce longer line length limit during "setup" phase of st…

### PR DESCRIPTION
**Pull request by poettering**
…ream protocol

Fixes: #18993
Replaces: #19008

### CODE REVIEW

**Comment by poettering** [comment]
@NickYang24 any chance you could test if this fixes #18993 for you?

File: src/journal/journald-stream.c [hunk 1]
```diff
@@ -39,6 +39,12 @@
 
 #define STDOUT_STREAMS_MAX 4096
 
+/* During the "setup" protocol phase of the stream logic let's define a different maximum line length than
+ * during the actual operational phase. We want to allow users to specify very short line lengths after all,
+ * but the unit name we embedd in the setup protocol might be longer than that. Hence, during the setup phase
```
Comment (bluca) [inline src/journal/journald-stream.c]: /embedd/embed/

**Review by bluca — APPROVED** [review]
LGTM - is this for 248 or later?

**Comment by bluca** [comment]
CI's not happy:

> Assertion 's->length <= limit' failed at src/journal/journald-stream.c:577, function int stdout_stream_process(sd_event_source *, int, uint32_t, void *)(). Aborting.

**Comment by poettering** [comment]
its for later

File: src/journal/journald-stream.c [hunk 2]
```diff
@@ -458,19 +464,33 @@ static int stdout_stream_found(
         return r;
 }
 
+static size_t stdout_stream_line_max(StdoutStream *s) {
+        assert(s);
+
+        /* During the "setup" phase of our protocol, let's ensure we we use a line length where a full unit
```
Comment (yuwata) [inline src/journal/journald-stream.c]: we we

**Comment by bluca** [comment]
This one doesn't look related:

> + cryptsetup close test-repart-25192
> Device test-repart-25192 is still in use.

**Comment by NickYang24** [comment]
/lgtm
It works, thank a lot.

**Comment by yuwata** [comment]
LGTM.

**Comment by NickYang24** [comment]
This pr seems introduced problems. As the log i listed below. If i set the LineMax to 10. The log of stdout shoule no longger than 79. But it seems there are some problems with this restrict.
Mar 23 15:09:06 localhost.localdomain bash[3526]: asdkfjasasdfjasdlfajssdjfsdjfskldfjsljfdklsjdflksjdflksjdflsjdflksjdflksdjflsdj
Mar 23 15:09:06 localhost.localdomain bash[3526]: flsdjfasdjfalsjdfalsdkjfasdjfaksldfjaaskldjfalskdjfasldjfasldjfasldfjalsdjfalsd
Mar 23 15:09:06 localhost.localdomain bash[3526]: jfasdlkfssssssssssssssssssssssssssssssssssssssssssssssssssssssaaaaaaaaaaaaaaaaa
Mar 23 15:09:06 localhost.localdomain bash[3526]: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaafffffffffffffffffffffffffffffffffffffffffffffffffffffffddddddddddddddddddddddddddddfdasjdfalsdjfaklsdjfasdfsdfhsdfhsdjfhsdkfhskj

**Comment by NickYang24** [comment]
I used journal -o verbose to get the specific log.
```
Tue 2021-03-23 15:05:54.980028 CST [s=3382e14c26794799860fc25c03afba7a;i=2e83;b=93714bd12abb4d89aedeb71d8640c86c;m=6df5319;t=5be2ed03cecbc;x=8bb106233aa073c]
    PRIORITY=6
    _BOOT_ID=93714bd12abb4d89aedeb71d8640c86c
    _MACHINE_ID=f4433d90df4f452da89b099fd40eec66
    _HOSTNAME=localhost.localdomain
    SYSLOG_FACILITY=3
    _UID=0
    _GID=0
    _SYSTEMD_SLICE=system.slice
    _TRANSPORT=stdout
    _LINE_BREAK=line-max
    _EXE=/usr/bin/bash
    _CAP_EFFECTIVE=3fffffffff
    _SELINUX_CONTEXT=system_u:system_r:init_t:s0
    _COMM=bash
    _PID=2255
    _CMDLINE=/usr/bin/bash /home/test.sh
    _SYSTEMD_CGROUP=/system.slice/test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfj>
    _SYSTEMD_UNIT=test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfjasldfsdjfasjdflk>
    _STREAM_ID=156be1ad60b841eab15b3ee6b7614c6e
    SYSLOG_IDENTIFIER=bash
    MESSAGE=jfasdlkfssssssssssssssssssssssssssssssssssssssssssssssssssssssaaaaaaaaaaaaaaaaa
Tue 2021-03-23 15:05:54.980608 CST [s=3382e14c26794799860fc25c03afba7a;i=2e84;b=93714bd12abb4d89aedeb71d8640c86c;m=6df555c;t=5be2ed03cef00;x=a42596ce120e59a9]
    PRIORITY=6
    _BOOT_ID=93714bd12abb4d89aedeb71d8640c86c
    _MACHINE_ID=f4433d90df4f452da89b099fd40eec66
    _HOSTNAME=localhost.localdomain
    SYSLOG_FACILITY=3
    _UID=0
    _GID=0
    _SYSTEMD_SLICE=system.slice
    _TRANSPORT=stdout
    _EXE=/usr/bin/bash
    _CAP_EFFECTIVE=3fffffffff
    _SELINUX_CONTEXT=system_u:system_r:init_t:s0
    _COMM=bash
    _PID=2255
    _CMDLINE=/usr/bin/bash /home/test.sh
    _SYSTEMD_CGROUP=/system.slice/test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfj>
    _SYSTEMD_UNIT=test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfjasldfsdjfasjdflk>
    _STREAM_ID=156be1ad60b841eab15b3ee6b7614c6e
    SYSLOG_IDENTIFIER=bash
    MESSAGE=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaafffffffffffffffffffffffffffffffffffffffffffffffffffffffdddddddddddddddddddddddddddd
```

**Comment by syyhao1994** [comment]
> 
> 
> I used journal -o verbose to get the specific log. So i think this pr needs to be reworked. And i tried the very beginning modifing that is ok with **limit = MIN(s->allocated - 1, stdout_stream_line_max(s));**, although the ci was not happy.
> 
> ```
> Tue 2021-03-23 15:05:54.980028 CST [s=3382e14c26794799860fc25c03afba7a;i=2e83;b=93714bd12abb4d89aedeb71d8640c86c;m=6df5319;t=5be2ed03cecbc;x=8bb106233aa073c]
>     PRIORITY=6
>     _BOOT_ID=93714bd12abb4d89aedeb71d8640c86c
>     _MACHINE_ID=f4433d90df4f452da89b099fd40eec66
>     _HOSTNAME=localhost.localdomain
>     SYSLOG_FACILITY=3
>     _UID=0
>     _GID=0
>     _SYSTEMD_SLICE=system.slice
>     _TRANSPORT=stdout
>     _LINE_BREAK=line-max
>     _EXE=/usr/bin/bash
>     _CAP_EFFECTIVE=3fffffffff
>     _SELINUX_CONTEXT=system_u:system_r:init_t:s0
>     _COMM=bash
>     _PID=2255
>     _CMDLINE=/usr/bin/bash /home/test.sh
>     _SYSTEMD_CGROUP=/system.slice/test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfj>
>     _SYSTEMD_UNIT=test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfjasldfsdjfasjdflk>
>     _STREAM_ID=156be1ad60b841eab15b3ee6b7614c6e
>     SYSLOG_IDENTIFIER=bash
>     MESSAGE=jfasdlkfssssssssssssssssssssssssssssssssssssssssssssssssssssssaaaaaaaaaaaaaaaaa
> Tue 2021-03-23 15:05:54.980608 CST [s=3382e14c26794799860fc25c03afba7a;i=2e84;b=93714bd12abb4d89aedeb71d8640c86c;m=6df555c;t=5be2ed03cef00;x=a42596ce120e59a9]
>     PRIORITY=6
>     _BOOT_ID=93714bd12abb4d89aedeb71d8640c86c
>     _MACHINE_ID=f4433d90df4f452da89b099fd40eec66
>     _HOSTNAME=localhost.localdomain
>     SYSLOG_FACILITY=3
>     _UID=0
>     _GID=0
>     _SYSTEMD_SLICE=system.slice
>     _TRANSPORT=stdout
>     _EXE=/usr/bin/bash
>     _CAP_EFFECTIVE=3fffffffff
>     _SELINUX_CONTEXT=system_u:system_r:init_t:s0
>     _COMM=bash
>     _PID=2255
>     _CMDLINE=/usr/bin/bash /home/test.sh
>     _SYSTEMD_CGROUP=/system.slice/test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfj>
>     _SYSTEMD_UNIT=test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfjasldfsdjfasjdflk>
>     _STREAM_ID=156be1ad60b841eab15b3ee6b7614c6e
>     SYSLOG_IDENTIFIER=bash
>     MESSAGE=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaafffffffffffffffffffffffffffffffffffffffffffffffffffffffdddddddddddddddddddddddddddd
> ```

I made a PR, could you try if it works for you please?
cc @poettering @bluca PTAL.
https://github.com/systemd/systemd/pull/19102/commits/1fe963b9c793b456157cc8495ac3374349c81a68

**Comment by syyhao1994** [comment]
> 
> 
> I used journal -o verbose to get the specific log. So i think this pr needs to be reworked. And i tried the very beginning modifing that is ok with **limit = MIN(s->allocated - 1, stdout_stream_line_max(s));**, although the ci was not happy.
> The test case i used is listed as below:
> 
> ```
> # cat /home/test.sh
> #!/bin/bash
> 
> echo asdkfjasasdfjasdlfajssdjfsdjfskldfjsljfdklsjdflksjdflksjdflsjdflksjdflksdjflsdjflsdjfasdjfalsjdfalsdkjfasdjfaksldfjaaskldjfalskdjfasldjfasldjfasldfjalsdjfalsdjfasdlkfssssssssssssssssssssssssssssssssssssssssssssssssssssssaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaafffffffffffffffffffffffffffffffffffffffffffffffffffffffdddddddddddddddddddddddddddd
> 
> # systemctl cat test.service
> # /usr/lib/systemd/system/test.service
> [Unit]
> Description=test journald line max
> 
> [Service]
> Type=oneshot
> ExecStart=/usr/bin/bash /home/test.sh
> ```
> 
> ```
> Tue 2021-03-23 15:05:54.980028 CST [s=3382e14c26794799860fc25c03afba7a;i=2e83;b=93714bd12abb4d89aedeb71d8640c86c;m=6df5319;t=5be2ed03cecbc;x=8bb106233aa073c]
>     PRIORITY=6
>     _BOOT_ID=93714bd12abb4d89aedeb71d8640c86c
>     _MACHINE_ID=f4433d90df4f452da89b099fd40eec66
>     _HOSTNAME=localhost.localdomain
>     SYSLOG_FACILITY=3
>     _UID=0
>     _GID=0
>     _SYSTEMD_SLICE=system.slice
>     _TRANSPORT=stdout
>     _LINE_BREAK=line-max
>     _EXE=/usr/bin/bash
>     _CAP_EFFECTIVE=3fffffffff
>     _SELINUX_CONTEXT=system_u:system_r:init_t:s0
>     _COMM=bash
>     _PID=2255
>     _CMDLINE=/usr/bin/bash /home/test.sh
>     _SYSTEMD_CGROUP=/system.slice/test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfj>
>     _SYSTEMD_UNIT=test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfjasldfsdjfasjdflk>
>     _STREAM_ID=156be1ad60b841eab15b3ee6b7614c6e
>     SYSLOG_IDENTIFIER=bash
>     MESSAGE=jfasdlkfssssssssssssssssssssssssssssssssssssssssssssssssssssssaaaaaaaaaaaaaaaaa
> Tue 2021-03-23 15:05:54.980608 CST [s=3382e14c26794799860fc25c03afba7a;i=2e84;b=93714bd12abb4d89aedeb71d8640c86c;m=6df555c;t=5be2ed03cef00;x=a42596ce120e59a9]
>     PRIORITY=6
>     _BOOT_ID=93714bd12abb4d89aedeb71d8640c86c
>     _MACHINE_ID=f4433d90df4f452da89b099fd40eec66
>     _HOSTNAME=localhost.localdomain
>     SYSLOG_FACILITY=3
>     _UID=0
>     _GID=0
>     _SYSTEMD_SLICE=system.slice
>     _TRANSPORT=stdout
>     _EXE=/usr/bin/bash
>     _CAP_EFFECTIVE=3fffffffff
>     _SELINUX_CONTEXT=system_u:system_r:init_t:s0
>     _COMM=bash
>     _PID=2255
>     _CMDLINE=/usr/bin/bash /home/test.sh
>     _SYSTEMD_CGROUP=/system.slice/test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfj>
>     _SYSTEMD_UNIT=test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfjasldfsdjfasjdflk>
>     _STREAM_ID=156be1ad60b841eab15b3ee6b7614c6e
>     SYSLOG_IDENTIFIER=bash
>     MESSAGE=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaafffffffffffffffffffffffffffffffffffffffffffffffffffffffdddddddddddddddddddddddddddd
> ```

I made a pr, could you try to test if this fixes your problem?
cc @poettering @bluca PTAL.
https://github.com/systemd/systemd/pull/19102

**Comment by keszybz** [comment]
@poettering could you take a look at @NickYang24's and @syyhao1994's comments above?

**Comment by NickYang24** [comment]
> 
> 
> > I used journal -o verbose to get the specific log. So i think this pr needs to be reworked. And i tried the very beginning modifing that is ok with **limit = MIN(s->allocated - 1, stdout_stream_line_max(s));**, although the ci was not happy.
> > The test case i used is listed as below:
> > ```
> > # cat /home/test.sh
> > #!/bin/bash
> > 
> > echo asdkfjasasdfjasdlfajssdjfsdjfskldfjsljfdklsjdflksjdflksjdflsjdflksjdflksdjflsdjflsdjfasdjfalsjdfalsdkjfasdjfaksldfjaaskldjfalskdjfasldjfasldjfasldfjalsdjfalsdjfasdlkfssssssssssssssssssssssssssssssssssssssssssssssssssssssaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaafffffffffffffffffffffffffffffffffffffffffffffffffffffffdddddddddddddddddddddddddddd
> > 
> > # systemctl cat test.service
> > # /usr/lib/systemd/system/test.service
> > [Unit]
> > Description=test journald line max
> > 
> > [Service]
> > Type=oneshot
> > ExecStart=/usr/bin/bash /home/test.sh
> > ```
> > 
> > 
> > ```
> > Tue 2021-03-23 15:05:54.980028 CST [s=3382e14c26794799860fc25c03afba7a;i=2e83;b=93714bd12abb4d89aedeb71d8640c86c;m=6df5319;t=5be2ed03cecbc;x=8bb106233aa073c]
> >     PRIORITY=6
> >     _BOOT_ID=93714bd12abb4d89aedeb71d8640c86c
> >     _MACHINE_ID=f4433d90df4f452da89b099fd40eec66
> >     _HOSTNAME=localhost.localdomain
> >     SYSLOG_FACILITY=3
> >     _UID=0
> >     _GID=0
> >     _SYSTEMD_SLICE=system.slice
> >     _TRANSPORT=stdout
> >     _LINE_BREAK=line-max
> >     _EXE=/usr/bin/bash
> >     _CAP_EFFECTIVE=3fffffffff
> >     _SELINUX_CONTEXT=system_u:system_r:init_t:s0
> >     _COMM=bash
> >     _PID=2255
> >     _CMDLINE=/usr/bin/bash /home/test.sh
> >     _SYSTEMD_CGROUP=/system.slice/test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfj>
> >     _SYSTEMD_UNIT=test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfjasldfsdjfasjdflk>
> >     _STREAM_ID=156be1ad60b841eab15b3ee6b7614c6e
> >     SYSLOG_IDENTIFIER=bash
> >     MESSAGE=jfasdlkfssssssssssssssssssssssssssssssssssssssssssssssssssssssaaaaaaaaaaaaaaaaa
> > Tue 2021-03-23 15:05:54.980608 CST [s=3382e14c26794799860fc25c03afba7a;i=2e84;b=93714bd12abb4d89aedeb71d8640c86c;m=6df555c;t=5be2ed03cef00;x=a42596ce120e59a9]
> >     PRIORITY=6
> >     _BOOT_ID=93714bd12abb4d89aedeb71d8640c86c
> >     _MACHINE_ID=f4433d90df4f452da89b099fd40eec66
> >     _HOSTNAME=localhost.localdomain
> >     SYSLOG_FACILITY=3
> >     _UID=0
> >     _GID=0
> >     _SYSTEMD_SLICE=system.slice
> >     _TRANSPORT=stdout
> >     _EXE=/usr/bin/bash
> >     _CAP_EFFECTIVE=3fffffffff
> >     _SELINUX_CONTEXT=system_u:system_r:init_t:s0
> >     _COMM=bash
> >     _PID=2255
> >     _CMDLINE=/usr/bin/bash /home/test.sh
> >     _SYSTEMD_CGROUP=/system.slice/test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfj>
> >     _SYSTEMD_UNIT=test1-asafsdfa-asdfjaslfjalsdjfalsdjfalsdfjalsdf-asdfjalsjfasldfjasldfjaskldfjaslkdfjalskdjfalsdjfaklsdjflkajsdflkasdflasdfasdfasdfasdfasdfaskdjfalskdfjalskjfalsjfalsffjasldfjasldfsdjfasjdflk>
> >     _STREAM_ID=156be1ad60b841eab15b3ee6b7614c6e
> >     SYSLOG_IDENTIFIER=bash
> >     MESSAGE=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaafffffffffffffffffffffffffffffffffffffffffffffffffffffffdddddddddddddddddddddddddddd
> > ```
> 
> I made a pr, could you try to test if this fixes your problem?
> cc @poettering @bluca PTAL.
> #19102

It seems works.
@poettering Could you take a look into this again?

**Comment by poettering** [comment]
Closing this one, let's continue in #19337

**Pull request closed without merging**